roachtest/sqlsmith: error decoding 10 bytes: empty array

Extracted from https://github.com/cockroachdb/cockroach/issues/70666#issuecomment-927242723.

roachtest.sqlsmith/setup=empty/setting=no-ddl [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3498637&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3498637&tab=artifacts#/sqlsmith/setup=empty/setting=no-ddl) on master @ [ecaa4920192549af1418923648769ede3de1c9c8](https://github.com/cockroachdb/cockroach/commits/ecaa4920192549af1418923648769ede3de1c9c8):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=empty/setting=no-ddl/run_1
	sqlsmith.go:223,sqlsmith.go:256,test_runner.go:777: error: pq: internal error: unexpected error from the vectorized engine: error decoding 10 bytes: empty array
		stmt:
SELECT
  col
FROM
  (VALUES ((ARRAY[]:::RECORD[],)), ((ARRAY[()],)))
    AS t(col)
ORDER BY
  col ASC;
```
<details><summary>Reproduce</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)

</p>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=empty/setting=no-ddl.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/distsql_physical_planner.go
+++ b/pkg/sql/distsql_physical_planner.go
@@ -248,12 +248,12 @@ func (v *distSQLExprCheckVisitor) VisitPre(expr tree.Expr) (recurse bool, newExp
 		}
 	case *tree.DArray:
 		// We need to check for arrays of untyped tuples here since constant-folding
-		// on builtin functions sometimes produces this.
+		// on builtin functions sometimes produces this. DecodeUntaggedDatum
+		// requires that all the types of the tuple contents are known.
 		if t.ResolvedType().ArrayContents() == types.AnyTuple {
 			v.err = newQueryNotSupportedErrorf("array %s cannot be executed with distsql", t)
 			return false, expr
 		}
-
 	}
 	return true, expr
 }

--- a/pkg/sql/logictest/testdata/logic_test/tuple
+++ b/pkg/sql/logictest/testdata/logic_test/tuple
@@ -979,3 +979,108 @@ UNION
 SELECT (1::INT, NULL)
 ----
 (1,)
+
+# Regression test for #70767. Make sure we avoid encoding arrays where the
+# array content type is AnyTuple.
+subtest regression_70767
+
+query T
+SELECT
+  col
+FROM
+  (VALUES ((ARRAY[]::RECORD[])), ((ARRAY[()])))
+    AS t(col)
+ORDER BY
+  col ASC
+----
+{}
+{()}
+
+query T
+SELECT
+  col
+FROM
+  (VALUES (NULL), ((ARRAY[]::RECORD[])), ((ARRAY[()])))
+    AS t(col)
+ORDER BY
+  col ASC
+----
+NULL
+{}
+{()}
+
+# ARRAY[(1)] is type-checked as an int[] -- this also matches Postgres.
+statement error VALUES types int\[\] and tuple\[\] cannot be matched
+SELECT
+  col
+FROM
+  (VALUES ((ARRAY[]::RECORD[])), ((ARRAY[(1)])))
+    AS t(col)
+ORDER BY
+  col ASC
+
+statement error VALUES types tuple{int, string}\[\] and tuple{int, int}\[\] cannot be matched
+SELECT
+  col
+FROM
+  (VALUES ((ARRAY[]::RECORD[])), ((ARRAY[(1,2)])), ((ARRAY[(1,'cat')])))
+    AS t(col)
+ORDER BY
+  col ASC
+
+query T
+SELECT
+  col
+FROM
+  (VALUES ((ARRAY[]::RECORD[])), ((ARRAY[(1,2)])))
+    AS t(col)
+ORDER BY
+  col ASC
+----
+{}
+{"(1,2)"}
+
+query T
+SELECT
+  col
+FROM
+  (VALUES ((ARRAY[(1,2)])), ((ARRAY[]::RECORD[])))
+    AS t(col)
+ORDER BY
+  col ASC
+----
+{}
+{"(1,2)"}
+
+query T
+SELECT
+  col
+FROM
+  (VALUES ((ARRAY[]::RECORD[])),  ((ARRAY[(1,2), (3,4)])))
+    AS t(col)
+ORDER BY
+  col ASC
+----
+{}
+{"(1,2)","(3,4)"}
+
+statement error expected tuple \(3, 4, 5\) to have a length of 2
+SELECT
+  col
+FROM
+  (VALUES ((ARRAY[]::RECORD[])),  ((ARRAY[(1,2), (3,4), (3,4,5)])))
+    AS t(col)
+ORDER BY
+  col ASC
+
+query T
+SELECT
+  col
+FROM
+  (VALUES ((ARRAY[]::RECORD[])), ((ARRAY[(1,'cat')])))
+    AS t(col)
+ORDER BY
+  col ASC
+----
+{}
+{"(1,cat)"}

--- a/pkg/sql/opt/optbuilder/values.go
+++ b/pkg/sql/opt/optbuilder/values.go
@@ -78,6 +78,15 @@ func (b *Builder) buildValuesClause(
 			if typ := texpr.ResolvedType(); typ.Family() != types.UnknownFamily {
 				if colTypes[colIdx].Family() == types.UnknownFamily {
 					colTypes[colIdx] = typ
+				} else if colTypes[colIdx].Family() == types.ArrayFamily &&
+					colTypes[colIdx].ArrayContents() == types.AnyTuple &&
+					typ.Family() == types.ArrayFamily &&
+					typ.ArrayContents().Family() == types.TupleFamily &&
+					typ.ArrayContents() != types.AnyTuple {
+					// This more complicated condition handles the case when an earlier
+					// expression in the VALUES clause is an array of AnyTuple, but a
+					// later expression is an array of a more specific tuple type.
+					colTypes[colIdx] = typ
 				} else if !typ.Equivalent(colTypes[colIdx]) {
 					panic(pgerror.Newf(pgcode.DatatypeMismatch,
 						"VALUES types %s and %s cannot be matched", typ, colTypes[colIdx]))

--- a/pkg/sql/sem/tree/values.go
+++ b/pkg/sql/sem/tree/values.go
@@ -36,15 +36,3 @@ func (node *ValuesClause) Format(ctx *FmtCtx) {
 		comma = ", "
 	}
 }
-
-// ValuesClauseWithNames is a VALUES clause that has been annotated with column
-// names. This is only produced at plan time, never by the parser. It's used to
-// pass column names to the VALUES planNode, so it can produce intelligible
-// error messages during value type checking.
-type ValuesClauseWithNames struct {
-	ValuesClause
-
-	// Names is a list of the column names that each tuple in the ValuesClause
-	// corresponds to.
-	Names NameList
-}

--- a/pkg/sql/values.go
+++ b/pkg/sql/values.go
@@ -12,15 +12,10 @@ package sql
 
 import (
 	"context"
-	"strconv"
 
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/colinfo"
-	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
-	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
 	"github.com/cockroachdb/cockroach/pkg/sql/rowcontainer"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
-	"github.com/cockroachdb/cockroach/pkg/sql/types"
-	"github.com/cockroachdb/errors"
 )
 
 type valuesNode struct {
@@ -35,82 +30,6 @@ type valuesNode struct {
 	specifiedInQuery bool
 
 	valuesRun
-}
-
-// Values implements the VALUES clause.
-func (p *planner) Values(
-	ctx context.Context, origN tree.Statement, desiredTypes []*types.T,
-) (planNode, error) {
-	v := &valuesNode{
-		specifiedInQuery: true,
-	}
-
-	// If we have names, extract them.
-	var n *tree.ValuesClause
-	switch t := origN.(type) {
-	case *tree.ValuesClauseWithNames:
-		n = &t.ValuesClause
-	case *tree.ValuesClause:
-		n = t
-	default:
-		return nil, errors.AssertionFailedf("unhandled case in values: %T %v", origN, origN)
-	}
-
-	if len(n.Rows) == 0 {
-		return v, nil
-	}
-
-	numCols := len(n.Rows[0])
-
-	v.tuples = make([][]tree.TypedExpr, 0, len(n.Rows))
-	tupleBuf := make([]tree.TypedExpr, len(n.Rows)*numCols)
-
-	v.columns = make(colinfo.ResultColumns, 0, numCols)
-
-	// We need to save and restore the previous value of the field in
-	// semaCtx in case we are recursively called within a subquery
-	// context.
-	defer p.semaCtx.Properties.Restore(p.semaCtx.Properties)
-
-	// Ensure there are no special functions in the clause.
-	p.semaCtx.Properties.Require("VALUES", tree.RejectSpecial)
-
-	for num, tuple := range n.Rows {
-		if a, e := len(tuple), numCols; a != e {
-			return nil, newValuesListLenErr(e, a)
-		}
-
-		// Chop off prefix of tupleBuf and limit its capacity.
-		tupleRow := tupleBuf[:numCols:numCols]
-		tupleBuf = tupleBuf[numCols:]
-
-		for i, expr := range tuple {
-			desired := types.Any
-			if len(desiredTypes) > i {
-				desired = desiredTypes[i]
-			}
-
-			// Clear the properties so we can check them below.
-			typedExpr, err := p.analyzeExpr(ctx, expr, nil, tree.IndexedVarHelper{}, desired, false, "")
-			if err != nil {
-				return nil, err
-			}
-
-			typ := typedExpr.ResolvedType()
-			if num == 0 {
-				v.columns = append(v.columns, colinfo.ResultColumn{Name: "column" + strconv.Itoa(i+1), Typ: typ})
-			} else if v.columns[i].Typ.Family() == types.UnknownFamily {
-				v.columns[i].Typ = typ
-			} else if typ.Family() != types.UnknownFamily && !typ.Equivalent(v.columns[i].Typ) {
-				return nil, pgerror.Newf(pgcode.DatatypeMismatch,
-					"VALUES types %s and %s cannot be matched", typ, v.columns[i].Typ)
-			}
-
-			tupleRow[i] = typedExpr
-		}
-		v.tuples = append(v.tuples, tupleRow)
-	}
-	return v, nil
 }
 
 func (p *planner) newContainerValuesNode(columns colinfo.ResultColumns, capacity int) *valuesNode {
@@ -183,11 +102,4 @@ func (n *valuesNode) Close(ctx context.Context) {
 		n.rows.Close(ctx)
 		n.rows = nil
 	}
-}
-
-func newValuesListLenErr(exp, got int) error {
-	return pgerror.Newf(
-		pgcode.Syntax,
-		"VALUES lists must all be the same length, expected %d columns, found %d",
-		exp, got)
 }

--- a/pkg/sql/values_test.go
+++ b/pkg/sql/values_test.go
@@ -11,180 +11,19 @@
 package sql
 
 import (
-	"context"
-	"fmt"
-	"go/constant"
-	"reflect"
 	"testing"
 	"time"
 
 	"github.com/cockroachdb/apd/v2"
-	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
-	"github.com/cockroachdb/cockroach/pkg/security"
-	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
-	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descs"
-	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
-	"github.com/cockroachdb/cockroach/pkg/sql/sessiondatapb"
 	"github.com/cockroachdb/cockroach/pkg/sql/types"
 	"github.com/cockroachdb/cockroach/pkg/util/bitarray"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
-	"github.com/cockroachdb/cockroach/pkg/util/mon"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
-	"github.com/cockroachdb/cockroach/pkg/util/uuid"
-	"github.com/cockroachdb/errors"
 	"github.com/stretchr/testify/require"
 )
-
-func makeTestPlanner() *planner {
-	// Initialize an Executorconfig sufficiently for the purposes of creating a
-	// planner.
-	settings := cluster.MakeTestingClusterSettings()
-	execCfg := ExecutorConfig{
-		Settings: settings,
-		NodeInfo: NodeInfo{
-			NodeID: base.TestingIDContainer,
-			ClusterID: func() uuid.UUID {
-				return uuid.MakeV4()
-			},
-		},
-		RootMemoryMonitor: mon.NewUnlimitedMonitor(context.Background(), "test", mon.MemoryResource, nil, nil, 0, nil),
-		CollectionFactory: descs.NewCollectionFactory(settings, nil, nil, nil),
-	}
-
-	// TODO(andrei): pass the cleanup along to the caller.
-	p, _ /* cleanup */ := NewInternalPlanner(
-		"test",
-		nil, /* txn */
-		security.RootUserName(),
-		&MemoryMetrics{},
-		&execCfg,
-		sessiondatapb.SessionData{},
-	)
-	return p.(*planner)
-}
-
-func TestValues(t *testing.T) {
-	defer leaktest.AfterTest(t)()
-	defer log.Scope(t).Close(t)
-
-	p := makeTestPlanner()
-
-	vInt := int64(5)
-	vNum := 3.14159
-	vStr := "two furs one cub"
-	vBool := true
-
-	unsupp := &tree.RangeCond{}
-
-	intVal := func(v int64) *tree.NumVal {
-		return tree.NewNumVal(
-			constant.MakeInt64(v),
-			"", /* origString */
-			false /* negative */)
-	}
-	floatVal := func(f float64) *tree.CastExpr {
-		return &tree.CastExpr{
-			Expr: tree.NewNumVal(
-				constant.MakeFloat64(f),
-				"", /* origString */
-				false /* negative */),
-			Type: types.Float,
-		}
-	}
-	asRow := func(datums ...tree.Datum) []tree.Datums {
-		return []tree.Datums{datums}
-	}
-
-	makeValues := func(tuples ...tree.Exprs) *tree.ValuesClause {
-		return &tree.ValuesClause{Rows: tuples}
-	}
-	makeTuple := func(exprs ...tree.Expr) tree.Exprs {
-		return tree.Exprs(exprs)
-	}
-
-	testCases := []struct {
-		stmt *tree.ValuesClause
-		rows []tree.Datums
-		ok   bool
-	}{
-		{
-			makeValues(makeTuple(intVal(vInt))),
-			asRow(tree.NewDInt(tree.DInt(vInt))),
-			true,
-		},
-		{
-			makeValues(makeTuple(intVal(vInt), intVal(vInt))),
-			asRow(tree.NewDInt(tree.DInt(vInt)), tree.NewDInt(tree.DInt(vInt))),
-			true,
-		},
-		{
-			makeValues(makeTuple(floatVal(vNum))),
-			asRow(tree.NewDFloat(tree.DFloat(vNum))),
-			true,
-		},
-		{
-			makeValues(makeTuple(tree.NewDString(vStr))),
-			asRow(tree.NewDString(vStr)),
-			true,
-		},
-		{
-			makeValues(makeTuple(tree.NewDBytes(tree.DBytes(vStr)))),
-			asRow(tree.NewDBytes(tree.DBytes(vStr))),
-			true,
-		},
-		{
-			makeValues(makeTuple(tree.MakeDBool(tree.DBool(vBool)))),
-			asRow(tree.MakeDBool(tree.DBool(vBool))),
-			true,
-		},
-		{
-			makeValues(makeTuple(unsupp)),
-			nil,
-			false,
-		},
-	}
-
-	ctx := context.Background()
-	for i, tc := range testCases {
-		t.Run(fmt.Sprintf("%d", i), func(t *testing.T) {
-			plan, err := func() (_ planNode, err error) {
-				defer func() {
-					if r := recover(); r != nil {
-						err = errors.Errorf("%v", r)
-					}
-				}()
-				return p.Values(context.Background(), tc.stmt, nil)
-			}()
-			if plan != nil {
-				defer plan.Close(ctx)
-			}
-			if err == nil != tc.ok {
-				t.Errorf("%d: error_expected=%t, but got error %v", i, tc.ok, err)
-			}
-			if plan == nil {
-				return
-			}
-			params := runParams{ctx: ctx, p: p, extendedEvalCtx: &p.extendedEvalCtx}
-			if err := startExec(params, plan); err != nil {
-				t.Fatalf("%d: unexpected error in Start: %v", i, err)
-			}
-			var rows []tree.Datums
-			next, err := plan.Next(params)
-			for ; next; next, err = plan.Next(params) {
-				rows = append(rows, plan.Values())
-			}
-			if err != nil {
-				t.Fatal(err)
-			}
-			if !reflect.DeepEqual(rows, tc.rows) {
-				t.Errorf("%d: expected rows:\n%+v\nactual rows:\n%+v", i, tc.rows, rows)
-			}
-		})
-	}
-}
 
 type floatAlias float32
 type boolAlias bool

